app/src/main/java/org/fossasia/susi/ai/skills/SkillsActivity.kt

```diff
@@ -127,13 +127,30 @@ class SkillsActivity : AppCompatActivity(), SkillFragmentCallback {
 
             edtSearch = action.customView.findViewById(R.id.edtSearch) as EditText //the text editor
 
-            //this is a listener to do a search when the user clicks on search button
-            edtSearch?.setOnEditorActionListener(TextView.OnEditorActionListener { v, actionId, event ->
-                if (actionId == EditorInfo.IME_ACTION_SEARCH) {
-                    doSearch((findViewById(R.id.edtSearch) as EditText).text.toString())
-                    return@OnEditorActionListener true
+            //this is a listener to do a search when users enters query in editText
+            edtSearch?.addTextChangedListener(object : TextWatcher {
+                /**
+                 * Variable used to remove unneccessary invoking of doSearch() method.
+                 */
+                var text: String = ""
+                var afterText: String = ""
+                var check = true
+                override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
+                    //no need to implement
+                }
+
+                override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
+                    //no need to implement
+                }
+
+                override fun afterTextChanged(p0: Editable?) {
+                    afterText = p0.toString()
+                    //checking that value exist in skills and if not exist comparing the length of the current text to the previous text where the skills are present
+                    if (check || (afterText.length <= text.length)) {
+                        check = doSearch(p0.toString())
+                        text = p0.toString()
```

Comment:
@codedsun How is this assignment helping? Will the 'text', being a local variable here, retain this value once the user enters another character in the edit search?
Would you please log the values of 'check' and 'text' and notice the changes, if any, in their values? I think the value of check will be true every time the text is modified in the search bar and it would be the same procedure happening as before. Please correct me if I am wrong, but, this is what I see.

Reply:
@arundhati24 It does work, the values of local variable are retained and as the case stated above, like you said when u enter arbitary string of infinite length, then also we were calling the doSearch function.

>Here's what I did, I assigned variables (text, afterText, check)
the check value is set to true for first time pass and text and afterText are empty strings.

For the above logic,
First pass 
>- check = true.
     afterText = text that user enters after every keyboard length, 
     and then doSearch on that text, if that returns true, we move to 2nd pass. else we got check to false and text = until where we were' getting the response to true.

So then for pass 2 if user types an extra length query for which no answer exist, check becomes false and we are saved from calling the function.
and when the user deletes or truncates the string to a shorter length which is less than or equal to the value of text we again call doSearch.


Example : Input Suneet and their is a skill named susi.

1. User enters 'S' check = true, doSearch return true, text = S (afterText = S)
2 User enters 'U' check = true, doSearch return true, text = SU (afterText = S)
3.User enters 'N' check = true, doSearch return false, text = 'SUN' (afterText='SUN')
4. Now user enters anything SUNEET, check has become to false and afterText.length > text.length

5. If user truncates some of the words, and like he typed the query SUNEET and then deletes and come backs to SUN, SU,S the doSearch function  is called again.